opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -2633,4 +2513,217 @@ namespace Opm
+        const unsigned waterCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::waterCompIdx);
+        // Correction salt rate; evaporated water does not contain salt
+        EvalWell cq_s_sm = cq_s[waterCompIdx] - vap_wat_rate;
+        if (this->isInjector()) {
+            cq_s_sm *= this->wsalt();
+        } else {
+            cq_s_sm *= this->extendEval(intQuants.fluidState().saltConcentration());
+        }
+
+        // Note. Efficiency factor is handled in the output layer
+        rate = cq_s_sm.value();
+
+        cq_s_sm *= this->well_efficiency_factor_;
+        return Base::restrictEval(cq_s_sm);
+    }
+
+
+    template <typename TypeTag>
+    typename StandardWell<TypeTag>::Eval
+    StandardWell<TypeTag>::
+    connectionRateEnergy(const double maxOilSaturation,
+                         const std::vector<EvalWell>& cq_s,
+                         const IntensiveQuantities& intQuants,
+                         DeferredLogger& deferred_logger) const
+    {
+        auto fs = intQuants.fluidState();
+        Eval result = 0;
+        for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx) {
+            if (!FluidSystem::phaseIsActive(phaseIdx)) {
+                continue;
+            }
+
+            // convert to reservoir conditions
+            EvalWell cq_r_thermal(this->primary_variables_.numWellEq() + Indices::numEq, 0.);
+            const unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::solventComponentIndex(phaseIdx));
+            const bool both_oil_gas = FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx);
+            if (!both_oil_gas || FluidSystem::waterPhaseIdx == phaseIdx) {
+                cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));
+            } else {
+                // remove dissolved gas and vapporized oil
+                const unsigned oilCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
+                const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
+                // q_os = q_or * b_o + rv * q_gr * b_g
+                // q_gs = q_gr * g_g + rs * q_or * b_o
+                // q_gr = 1 / (b_g * d) * (q_gs - rs * q_os)
+                // d = 1.0 - rs * rv
+                const EvalWell d = this->extendEval(1.0 - fs.Rv() * fs.Rs());
+                if (d <= 0.0) {
+                    deferred_logger.debug(
+                        fmt::format("Problematic d value {} obtained for well {}"
+                                    " during calculateSinglePerf with rs {}"
+                                    ", rv {}. Continue as if no dissolution (rs = 0) and"
+                                    " vaporization (rv = 0) for this connection.",
+                                    d, this->name(), fs.Rs(), fs.Rv()));
```

Comment:
there are multiple uses of {fmt} in StandardWell and Multisegment well already so it's nothing new. I think the issue you think of is in opm-common headers (where there is a no-use-in-headers requirement), since opm-grid does not have an embedded libfmt.

Reply:
Wow,l always surprising this embedded libraries. Indded we have libfmt embedded in two OPM modules (opm-common and opm-simulators) :see_no_evil: